File: src/python/finished/maximum_number_of_words_you_can_type.py
# https://leetcode.com/problems/maximum-number-of-words-you-can-type/


class Solution:
    def canBeTypedWords(self, text: str, brokenLetters: str) -> int:
        words = text.split(" ")
        return len(words) - sum(
            1 if any(ch in brokenLetters for ch in w) else 0 for w in words
        )


class Solution:
    def canBeTypedWords(self, text: str, brokenLetters: str) -> int:
        words = text.split(" ")
        word_masks = [0] * len(words)

        for i, w in enumerate(words):
            mask = 0
            for ch in w:
                mask |= 1 << (ord(ch) - ord("a"))
            word_masks[i] = mask

        bad_mask = 0
        for ch in brokenLetters:
            bad_mask |= 1 << (ord(ch) - ord("a"))

        return sum(0 if (bad_mask & word_mask) else 1 for word_mask in word_masks)

File: src/rust/finished/maximum_number_of_words_you_can_type.rs
impl Solution {
  pub fn can_be_typed_words(text: String, broken_letters: String) -> i32 {
      let bad_mask = broken_letters.chars().fold(0_u32, |prev, cur| {
          prev | 1 << (cur as u8) as u32
      });

      text
          .split_ascii_whitespace()
          .map(|w| w.chars().fold(0_u32, |prev, cur| {
              prev | 1 << (cur as u8) as u32
          }))
          .map(|mask| {
              match mask & bad_mask {
                  0 => 1,
                  _ => 0,
              }
          })
          .sum()
  }
}